### PR TITLE
Reduce number of requests in scheduledSSTTimeSeriesUpdate

### PR DESCRIPTION
**Pull request by echaidemenos**
The purpose of this PR is to solve an issue where not all sites would get SST updates daily from the functions.
We suspect that sofar API fails to answer the big amount of the functions' requests.
This PR combines the 4 per site requests to 1 per site, by combining the start and end API parameters.
It also cleans up the functions logs so they are more useful.

Also since the sofar hindcast gives only one value per day, we could reduce the `scheduledSSTTimeSeriesUpdate` function calls from 1 per hour to one per day.

### CODE REVIEW

File: packages/api/src/utils/sst-time-series.ts [hunk 1]
```diff
@@ -82,89 +82,105 @@ export const updateSST = async (
...
-              getLatestData(sstFiltered)?.value,
-              // Calculate degree heating days
-              latestDhw && latestDhw.value * 7,
-            );
-
-            // Calculate the sstAnomaly
-            const sstAnomaly = sstFiltered
+      const endDate = moment().format();
+      const startDate = moment()
+        .subtract(days - 1, 'd')
```
Comment (ericboucher) [inline packages/api/src/utils/sst-time-series.ts:87]: I think we can take:
- `.endOf('day)` for the endDate and
- and simply `.subtract(days, 'd')` for startDate

File: packages/api/src/utils/sst-time-series.ts [hunk 2]
```diff
@@ -82,89 +82,105 @@ export const updateSST = async (
...
-              site.maxMonthlyMean,
-              getLatestData(sstFiltered)?.value,
-              // Calculate degree heating days
-              latestDhw && latestDhw.value * 7,
-            );
-
-            // Calculate the sstAnomaly
-            const sstAnomaly = sstFiltered
+      const endDate = moment().format();
+      const startDate = moment()
```
Comment (ericboucher) [inline packages/api/src/utils/sst-time-series.ts:86]: Is there a limit at how many days the SofarAPI will return?

File: packages/api/src/utils/sst-time-series.ts [hunk 3]
```diff
@@ -82,89 +82,105 @@ export const updateSST = async (
...
+        alert: alertLevel,
+      };
+
+      if (!data.sst.length || data.sst.length === 0) {
+        console.error(`No Hindcast data available for site '${site.id}'`);
+      }
+
       return Bluebird.map(
-        data,
+        [data],
```
Comment (ericboucher) [inline packages/api/src/utils/sst-time-series.ts:183]: Do we still need the `bluebird.map` this change makes it feel like we don't?

File: packages/api/scripts/backfill-sofar-time-series.ts [hunk 4]
```diff
@@ -84,7 +84,6 @@ async function run() {
   return fn(
     parsedSiteIds,
     days,
-    connection,
```
Comment (ericboucher) [inline packages/api/scripts/backfill-sofar-time-series.ts:87]: Why this change?

**Review by ericboucher — APPROVED** [review]
Looks great overall, thanks for looking into this! I left a few small comments for potential improvements inline

File: packages/api/src/utils/sst-time-series.ts (hunk 1 above)
Comment (echaidemenos) [reply packages/api/src/utils/sst-time-series.ts:87]: `endDate` must be in the past, so `.endOf('day')` wont work.
Ok about `startDate`.

File: packages/api/src/utils/sst-time-series.ts (hunk 2 above)
Comment (echaidemenos) [reply packages/api/src/utils/sst-time-series.ts:86]: turns out that there is a 7 day limit. I will have to keep some of the old logic to take that into account for the backfill script. The function will run OK since it only need 4 days

File: packages/api/src/utils/sst-time-series.ts [hunk 5]
```diff
@@ -82,89 +82,105 @@ export const updateSST = async (
...
       );
 
+      const data = {
+        sst: sstFiltered,
+        dhw: dhwFiltered,
+        sstAnomaly: anomaly,
+        alert: alertLevel,
+      };
+
+      if (!data.sst.length || data.sst.length === 0) {
```
[the remark this answers is not in the document]
Comment (echaidemenos) [reply packages/api/src/utils/sst-time-series.ts:178]: `sofarHindcast` already does error handling. This is in case the values are `null`, which is the case of the point not being in the availability zone.

File: packages/api/scripts/backfill-sofar-time-series.ts (hunk 4 above)
Comment (echaidemenos) [reply packages/api/scripts/backfill-sofar-time-series.ts:87]: It was left there unneeded form a previous PR. It is not being used so I removed it

File: packages/api/src/utils/sst-time-series.ts [hunk 6]
```diff
@@ -82,89 +82,105 @@ export const updateSST = async (
...
+            site.maxMonthlyMean,
+            latest?.value,
+            // Calculate degree heating days
+            dhw && dhw.value * 7,
+          );
+          if (!alert) return undefined;
+          return {
+            value: alert,
+            timestamp: latest?.timestamp,
+          } as ValueWithTimestamp;
```
[the remark this answers is not in the document]
Comment (echaidemenos) [reply packages/api/src/utils/sst-time-series.ts]: Thanks a lot for the detailed comment! You are right setting `if (!latest) return undefined;` is the better approach here.
`if (!alert) return undefined;` actually implies also `latest` not being nullish, since they are linked by the logic above

File: packages/api/src/utils/sst-time-series.ts [hunk 7]
```diff
@@ -82,89 +82,105 @@ export const updateSST = async (
...
+          startDate,
+          endDate,
+        ),
+      ]);
+
+      // Filter out null values
+      const sstFiltered = filterSofarResponse(SofarSSTRaw);
+      const dhwFiltered = filterSofarResponse(sofarDegreeHeatingWeekRaw);
+
+      const getDateNoTime = (x?: string) => new Date(x || '').toDateString();
```
[the remark this answers is not in the document]
Comment (echaidemenos) [reply packages/api/src/utils/sst-time-series.ts]: Thanks for the comment. Yes this is the desired behavior. `Invalid Date` is still a `Date` object and `toDateString` it's function. 
Although since you noted that i noticed that there is a chance `groupedDHWFiltered` and `groupedSSTFiltered` getting the key `Invalid Date`. I will make sure to have that key and its properties removed.

File: packages/api/src/utils/sst-time-series.ts (hunk 1 above)
Comment (ericboucher) [reply packages/api/src/utils/sst-time-series.ts]: Ah I see, I thought endDate could be an input as well. Then it makes sense.
